pkg/frontend/querymiddleware/cardinality.go

```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/grafana/dskit/cache"
+	"github.com/grafana/dskit/tenant"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+const (
+	// cardinalityEstimateTTL is how long a key must see no write to expire and be removed from the cache.
+	cardinalityEstimateTTL = 7 * 24 * time.Hour
+
+	// cardinalityEstimateBucketSize is the size of buckets that queries are bucketed into.
+	cardinalityEstimateBucketSize = 24 * time.Hour
+
+	// cacheErrorToleranceFraction is how much the estimate must deviate
+	// from the actually observed cardinality to update the cache.
+	cacheErrorToleranceFraction = 0.1
+)
+
+type cardinalityEstimationMetrics struct {
+	estimationError prometheus.Histogram
+}
+
+// cardinalityEstimation is a Handler that caches estimates for a query's
+// cardinality based on similar queries seen previously.
+type cardinalityEstimation struct {
+	cache  cache.Cache
+	next   Handler
+	logger log.Logger
+
+	cardinalityEstimationMetrics
+}
+
+func newCardinalityEstimationMiddleware(cache cache.Cache, logger log.Logger, registerer prometheus.Registerer) Middleware {
+	metrics := cardinalityEstimationMetrics{estimationError: promauto.With(registerer).NewHistogram(prometheus.HistogramOpts{
+		Name:    "cortex_frontend_cardinality_estimation_error",
```

Comment:
[design] "Error" is typically used for... errors. Try to come up with a better naming which doesn't use the "error" name.

Reply:
I think `error` is the scientifically correct term for this kind of difference (e.g. [wiki](https://en.wikipedia.org/wiki/Approximation_error), but if we want to avoid the `error` word here we could maybe use `cortex_frontend_cardinality_estimation_error`, WDYT?

Reply:
> cortex_frontend_cardinality_estimation_difference

LGTM. Also please use `cortex_query_frontend` prefix. The `cortex_frontend` metric name prefix is a legacy one.